text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,267 @@
+
+### `lwc:index`
+
+Identical to `for:index`. Returns a number for the current list index.
+
+As with `for:index`, `lwc:index` cannot use the same string as the `lwc:item` on the same element:
+
+```html
+<template lwc:each={items} lwc:key={same.id} lwc:item="same" lwc:index="same">
+</template>
+```
+
+The above will throw a compile-time error.
+
+(Note that `for:index` only [incidentally](https://stackblitz.com/edit/salesforce-lwc-se51rr?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground) throws an error for the same case, because the `key` ends up referencing the `for:index` variable. For `lwc:index`, the compile-time error should be explicit.)
+
+### `lwc:first`, `lwc:last`, `lwc:even`, `lwc:odd`
+
+These new directives take strings as their values. When referenced, they resolve to a boolean representing whether the list item is first, last, even, or odd. (This is heavily inspired by [Angular](https://angular.io/api/common/NgForOf#local-variables).)
+
+The addition of `lwc:first` and `lwc:last` serve to unify the two existing directives – previously, `iterator:*` was the only way to get "first" or "last" behavior. The addition of `lwc:even` and `lwc:odd` are merely for convenience (especially in the absence of [complex template expressions](https://github.com/salesforce/lwc/pull/3376)).
+
+### Constraints
+
+All new directives that take a string as their value (`lwc:item`, `lwc:index`, `lwc:first`, `lwc:last`, `lwc:even`, and `lwc:odd`) must have unique values when used on the same element.
+
+For example, the following will throw a compile-time error, because `lwc:item` and `lwc:index` have the same values:
+
+```html
+<template lwc:each={items} lwc:key={same.id} lwc:item="same" lwc:index="same">
+</template>
+```
+
+Empty values (e.g. `lwc:index=""`) will also throw a compile-time error. For example:
+
+```html
+<template lwc:each={items} lwc:key={item.id} lwc:item="item" lwc:index="">
+</template>
+```
+
+Missing values (e.g. `lwc:index`) will also throw at compile-time:
+
+```html
+<template lwc:each={items} lwc:key={item.id} lwc:item="item" lwc:index>
+</template>
+```
+
+The directives may appear in any order on the element.
+
+Any other constraints that apply to `for:each` must also apply to `lwc:each`. For example, `lwc:ref` is (currently) unsupported inside of `for:each` and `iterator:*` – the same applies to `lwc:each`. For another example, `<slot>`s cannot have any of the new directives, and cannot be repeated inside the new iterator.
+
+## Drawbacks
+
+This new directive may cause confusion, because we are introducing a third directive to try to unify the previous two (see [XKCD 927](https://xkcd.com/927/)).
+
+The implementation also becomes more complex, as we will need to support all three directives for the foreseeable future. Perhaps they can share logic under the hood, but we will still need to test all three, at the very least.

```

Comment:
Should the usage of `for:each` and `iterator:it` prohibit if a template has `lwc:each` usage? 
1. This will simplify our constraints by eliminating these mixed use cases.
2. Provide a path to deprecate the older forms. Will make sure for:each is truly a union of the prior two forms.

Reply:
Interesting, do you mean to say that `lwc:each` should throw a compiler error if it is mixed in the same file with either `for:each` or `iterator:*`?

I'm perfectly okay with this constraint. It would certainly make it simpler for us to write tests for loops-within-loops.

Reply:
> I'm also okay with extending this restriction to any directive associated with for:each and iterator:* – for example, key and for:item.

If we separate usage of `lwc:each` and `for:each/iterator:*` at the template level, should we remove the [warnings](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/template-compiler/src/parser/index.ts#L1328)/[restrictions](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/template-compiler/src/parser/index.ts#L1300) on the `key` directive so component authors can choose to use a `key` attribute on their elements?

This might cause confusion when it can be used though.

Reply:
> I'm also okay with extending this restriction to any directive associated with for:each and iterator:* – for example, key and for:item.

I'm in favor of restricting the directives associated with `for:each` and `iterator:*` as well to prevent confusion, especially around when `key` is available to use.

Reply:
4dbf178

Reply:
> Interesting, do you mean to say that lwc:each should throw a compiler error if it is mixed in the same file with either for:each or iterator:*?

Yes.